[clang][AArch64] Add getHostCPUFeatures to query for enabled features in cpu info

neildhickey opened the pull request:
This pull request adds code to call getHostCPUInfo into the AArch64 Target Parser to query the cpuinfo for the device in the case where we are compiling with -mcpu=native

neildhickey committed:
[AArch64] Add getHostCPUFeatures to query for enabled features in cpuinfo

davemgreen commented:
Hi this sounds like a good idea to me. Note that the implementation of getHostCPUFeatures isn't amazing for AArch64 at the moment, there was an attempt to fix it up in #95694 (but thaat has gone a bit quiet). One point we noticed is that it could end up turning "aes+sha2" into "crypto" and "crypto" back into "sha2+aes+sha3+sm4", as it uses the old meaning of "crypto"

neildhickey commented:
Thanks for the review comments @madhur13490 and @davemgreen I plan to improve getHostCPUFeatures for AArach64 as well, this is just the first phase to add the function call.

DavidSpickett commented:
Looks ok to me, X86 and ARM already do this.

> One point we noticed is that it could end up turning "aes+sha2" into "crypto" and "crypto" back into "sha2+aes+sha3+sm4", as it uses the old meaning of "crypto"

Then this needs to be fixed before this PR can go in, we don't want to be making binaries folks can't run. @davemgreen do we have a way that @neildhickey can reproduce that easily?

tmatheson-arm commented on clang/lib/Driver/ToolChains/Arch/AArch64.cpp:
```diff
@@ -445,4 +445,21 @@ void aarch64::getAArch64TargetFeatures(const Driver &D,
+    if (CPUName == "native") {
+      llvm::StringMap<bool> HostFeatures;
+      if (llvm::sys::getHostCPUFeatures(HostFeatures)) {
+        for (auto &F : HostFeatures) {
+          Features.push_back(
+            Args.MakeArgString((F.second ? "+" : "-") + F.first()));
+        }
```
```suggestion
        for (auto &[Name, Enabled] : HostFeatures) {
          Features.push_back(
            Args.MakeArgString((Enabled ? "+" : "-") + Name));
        }
```

tmatheson-arm commented on clang/lib/Driver/ToolChains/Arch/AArch64.cpp:
```diff
@@ -445,4 +445,21 @@ void aarch64::getAArch64TargetFeatures(const Driver &D,
     Features.push_back("+no-bti-at-return-twice");
+
+  // Parse AArch64 CPU Features
+  const Arg *CPUArg = Args.getLastArg(options::OPT_mcpu_EQ);
```
Feels like this should be done in `getAArch64ArchFeaturesFromMcpu`/`DecodeAArch64Mcpu`

neildhickey commented:
> Looks ok to me, X86 and ARM already do this.
> 
> > One point we noticed is that it could end up turning "aes+sha2" into "crypto" and "crypto" back into "sha2+aes+sha3+sm4", as it uses the old meaning of "crypto"
> 
> Then this needs to be fixed before this PR can go in, we don't want to be making binaries folks can't run. @davemgreen do we have a way that @neildhickey can reproduce that easily?

Perhaps this is the part of code causing the issue
```
#if defined(__aarch64__)
  // If we have all crypto bits we can add the feature
  if (crypto == (CAP_AES | CAP_PMULL | CAP_SHA1 | CAP_SHA2))
    Features["crypto"] = true;
#endif
```

If so I can change this to remove the addition of crypto and just assert the features found in the cpuinfo file

davemgreen commented:
It is that bit of code, yeah. I don't know of a way to reproduce this without logging into different machines with different sets of options and trying it.

If we had a way to test/mock that various /proc/cpuinfo files gave us the correct results, that would be helpful in giving us more confidence it was working as intended.

ElvinaYakubova committed:
[clang][AArch64] Add LLVM_CPUINFO environment variable

Add LLVM_CPUINFO environment variable to test mock /proc/cpuinfo
files for -mcpu=native

davemgreen commented on llvm/lib/TargetParser/Host.cpp:
```diff
@@ -70,6 +70,10 @@ static std::unique_ptr<llvm::MemoryBuffer>
     LLVM_ATTRIBUTE_UNUSED getProcCpuinfoContent() {
   llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Text =
       llvm::MemoryBuffer::getFileAsStream("/proc/cpuinfo");
+  if (const char *cpuinfoIntercept = std::getenv("LLVM_CPUINFO")) {
```
cpuinfoIntercept -> CpuinfoIntercept

This might be a little better to only open a single file
```
const char *CPUInfoFile = "/proc/cpuinfo";
if (const char *CpuinfoIntercept = std::getenv("LLVM_CPUINFO"))
  CPUInfoFile = CpuinfoIntercept;
llvm::ErrorOr<std::unique_ptr<llvm::MemoryBuffer>> Text =
      llvm::MemoryBuffer::getFileAsStream(CPUInfoFile);
```

ElvinaYakubova replied:
Thanks, changed that

davemgreen commented on llvm/lib/TargetParser/Host.cpp:
```diff
@@ -70,6 +70,10 @@ static std::unique_ptr<llvm::MemoryBuffer>
+
   if (std::error_code EC = Text.getError()) {
     llvm::errs() << "Can't read "
                  << "/proc/cpuinfo: " << EC.message() << "\n";
```
This error message could then use the filename as above.

davemgreen reviewed (commented):
Thanks for adding the tests. They look like a useful addition.

ElvinaYakubova commented:
Sorry for the delays in response. I addressed the comments and rebased the branch

davemgreen commented on clang/test/Driver/aarch64-mcpu-native.c:
```diff
@@ -0,0 +1,137 @@
+// RUN: export LLVM_CPUINFO=%S/Inputs/cpunative/neoverse-v2
+// RUN: %clang --target=aarch64 --print-enabled-extensions -mcpu=native | FileCheck --strict-whitespace --check-prefix=CHECK-FEAT-NV2 --implicit-check-not=FEAT_ %s
+
```
Will this test need `// REQUIRES: aarch64-registered-target`?

ElvinaYakubova replied:
thanks, you're right, added it

davemgreen reviewed (approved):
Thanks - LGTM
